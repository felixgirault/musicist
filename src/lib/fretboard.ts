import {sum} from './arrays';
import {
	type Note,
	type PitchClass,
	PitchClasses,
	notePitchClass
} from './notes';
import type {Scale} from './scales';

export type Fretboard = {
	tuning: Note[];
	fretCount: number;
};

// Magic constant representing the relationship in size
// between two frets.
// This is not an accurate value but it is more visually
// pleasing than the real one.
const FretRatio = 1.055;

// Builds a list of fret sizes as percentages, so they all
// add up to 100%.
export const fretRatios = (count: number) => {
	const ratios = Array(count)
		.fill(0)
		.reduce(
			(r) => r.concat((r?.[r.length - 1] ?? 1) / FretRatio),
			[] as number[]
		);

	const total = sum(ratios);
	return ratios.map((r) => (r / total) * 100);
};

// Finds a playable fingering for the given scale, ensuring
// there is not too much notes per string, and there is no
// excessive finger stretch between notes.
export const fingering = (
	fretboard: Fretboard,
	scale: Scale,
	root: PitchClass,
	{
		maxNotesPerString = 4,
		maxStretch = 4
	}: {
		maxNotesPerString?: number;
		maxStretch?: number;
	} = {}
) => {
	// In the rest of the function, PCI stands for "pitch
	// class index", to shorten variable names a bit.
	// Everything is computed by comparing the index of pitch
	// classes in the full pitch class list, as to speed up
	// computations by avoiding comparing notes, transposing
	// them, etc.
	const rootPci = PitchClasses.indexOf(root);
	const openPcis = fretboard.tuning.map((note) =>
		PitchClasses.indexOf(notePitchClass(note))
	);

	const scalePcis = scale.chroma
		.split('')
		.reduce(
			(pcis, chroma, i) =>
				chroma === '1'
					? pcis.concat((rootPci + i) % PitchClasses.length)
					: pcis,
			[] as number[]
		);

	const indices: number[][] = openPcis.map(() => []);

	let nextScalePciIndex = 0;
	let rootFretIndex = -1;

	for (
		let stringIndex = 0;
		stringIndex < openPcis.length;
		stringIndex++
	) {
		const openPci = openPcis[stringIndex];
		let pressedFretCount = 0;

		for (
			let fretIndex = 0;
			fretIndex < fretboard.fretCount;
			fretIndex++
		) {
			const currentPci =
				(openPci + fretIndex) % PitchClasses.length;
			const nextScalePci = scalePcis[nextScalePciIndex];

			// Skips notes until the root is found.
			if (rootFretIndex < 0) {
				if (currentPci === nextScalePci) {
					rootFretIndex = fretIndex;
				} else {
					continue;
				}
			}

			if (currentPci !== nextScalePci) {
				continue;
			}

			// Is the finger stretch required to play this note
			// too wide?
			if (Math.abs(rootFretIndex - fretIndex) > maxStretch) {
				break;
			}

			pressedFretCount++;
			nextScalePciIndex =
				(nextScalePciIndex + 1) % scalePcis.length;
			indices[stringIndex].push(fretIndex);

			if (pressedFretCount >= maxNotesPerString) {
				break;
			}
		}
	}

	return indices;
};
