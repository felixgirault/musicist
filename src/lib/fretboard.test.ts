import {ScaleType} from '@tonaljs/tonal';
import {describe, expect, test} from 'vitest';
import {fingering} from './fretboard';
import type {Note, PitchClass} from './notes';

describe('fingering', () => {
	const fretboard = {
		tuning: ['E2', 'A2', 'D3', 'G3', 'B3', 'E4'] as Note[],
		fretCount: 12
	};

	test.each<[PitchClass, string]>([
		['A', 'chromatic'],
		['A', 'flamenco'],
		['A', 'major blues'],
		['A', 'minor pentatonic']
	])('fingering for %s %s', (root, scale) => {
		expect(
			fingering(fretboard, ScaleType.get(scale), root)
		).toMatchSnapshot();
	});
});
