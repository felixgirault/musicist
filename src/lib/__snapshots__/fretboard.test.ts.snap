// Vitest Snapshot v1

exports[`fingering > fingering for A chromatic 1`] = `
[
  [
    5,
    6,
    7,
    8,
  ],
  [
    4,
    5,
    6,
    7,
  ],
  [
    3,
    4,
    5,
    6,
  ],
  [
    2,
    3,
    4,
    5,
  ],
  [
    2,
    3,
    4,
    5,
  ],
  [
    1,
    2,
    3,
    4,
  ],
]
`;

exports[`fingering > fingering for A flamenco 1`] = `
[
  [
    5,
    6,
    8,
    9,
  ],
  [
    6,
    7,
  ],
  [
    5,
    7,
    8,
  ],
  [
    5,
    6,
    8,
    9,
  ],
  [
    8,
  ],
  [
    5,
    6,
    8,
    9,
  ],
]
`;

exports[`fingering > fingering for A major blues 1`] = `
[
  [
    5,
    7,
    8,
    9,
  ],
  [
    7,
    9,
  ],
  [
    7,
    9,
  ],
  [
    5,
    6,
    9,
  ],
  [
    7,
  ],
  [
    5,
    7,
    8,
    9,
  ],
]
`;

exports[`fingering > fingering for A minor pentatonic 1`] = `
[
  [
    5,
    8,
  ],
  [
    5,
    7,
  ],
  [
    5,
    7,
  ],
  [
    5,
    7,
    9,
  ],
  [
    8,
  ],
  [
    5,
    8,
  ],
]
`;
